File: src/macos/LanMessenger/Core/Networking/Media/MediaReassembler.swift
import Foundation

// The demux: per-channel fragment buffers, so an interleaved control or input
// frame arriving between two video segments never disturbs the partially
// assembled video frame. That per-channel keying is the reader-side half of the
// writer's interleaving design — without it, interleaving would corrupt video.

/// A fully reassembled inbound frame.
struct MediaInboundFrame: Equatable {
    let channel: MediaChannel
    let flags: MediaFlags
    /// Sequence of the FINAL fragment.
    let sequence: UInt64
    /// Capture timestamp from the FIRST fragment — the whole frame was captured
    /// at one instant, and later fragments carry the same value only by
    /// convention, so the first is the authoritative one.
    let captureUs: UInt64
    let payload: Data
    let fragmentCount: Int
}

final class MediaReassembler {

    enum Outcome: Equatable {
        case complete(MediaInboundFrame)
        case partial
        case fault(MediaProtocolError)
    }

    private struct Pending {
        var payload: Data
        var captureUs: UInt64
        var flags: MediaFlags
        var fragmentCount: Int
    }

    private var pending: [UInt8: Pending] = [:]
    private let maxAssembledBytes: Int
    private let maxFragments: Int

    /// Two caps, not one, and they close different holes.
    ///
    /// The 4 MiB wire cap bounds a single *frame*; it says nothing about the
    /// reassembled total, so the byte cap is needed to bound a long fragment
    /// chain. But a byte cap alone still lets a stream of zero-length fragments
    /// grow a per-channel buffer forever while every individual frame passes the
    /// wire check — a real allocation vector — so the fragment count is capped
    /// independently.
    init(maxAssembledBytes: Int = MediaFrameCodec.maxFrameLength, maxFragments: Int = 512) {
        self.maxAssembledBytes = maxAssembledBytes
        self.maxFragments = maxFragments
    }

    func accept(header: MediaFrameHeader, plaintext: Data) -> Outcome {
        guard let channel = header.channel else {
            // Callers filter unknown channels before here; treat defensively.
            return .partial
        }
        let key = header.rawChannel

        // Not fragmented: a whole frame in one go.
        guard header.isFragmented else {
            if pending[key] != nil {
                // The writer never does this, so it is corruption or an attack.
                return .fault(.wholeFrameMidReassembly(channel))
            }
            return .complete(MediaInboundFrame(
                channel: channel, flags: header.flags, sequence: header.sequence,
                captureUs: header.captureUs, payload: plaintext, fragmentCount: 1))
        }

        var entry = pending[key] ?? Pending(
            payload: Data(), captureUs: header.captureUs, flags: header.flags, fragmentCount: 0)

        guard entry.fragmentCount + 1 <= maxFragments else {
            pending[key] = nil
            return .fault(.fragmentCountExceeded(channel, count: entry.fragmentCount + 1))
        }
        guard entry.payload.count + plaintext.count <= maxAssembledBytes else {
            pending[key] = nil
            return .fault(.reassemblyOverflow(channel, bytes: entry.payload.count + plaintext.count))
        }

        entry.payload.append(plaintext)
        entry.fragmentCount += 1

        guard header.isFinalFragment else {
            pending[key] = entry
            return .partial
        }

        pending[key] = nil
        return .complete(MediaInboundFrame(
            channel: channel,
            flags: entry.flags,          // first fragment's flags, incl. keyframe
            sequence: header.sequence,   // final fragment's sequence
            captureUs: entry.captureUs,  // first fragment's timestamp
            payload: entry.payload,
            fragmentCount: entry.fragmentCount))
    }

    /// Bytes currently held for a channel. Tests and the stats summary only.
    func pendingBytes(for channel: MediaChannel) -> Int {
        pending[channel.rawValue]?.payload.count ?? 0
    }

    func reset() { pending.removeAll() }
}

// MARK: - Inbound sequence enforcement

/// Strictly-increasing sequence enforcement.
///
/// `lastAccepted` starts nil rather than 0, and that choice is load-bearing.
/// The protocol says a reconnect restarts sequencing from zero against a key
/// that has never been used, so the first frame of a session legitimately
/// carries sequence 0. Initialising to 0 and requiring `> last` would reject it,
/// and the session would die at `hello` with a sequence violation that looks
/// exactly like a crypto fault — a miserable thing to diagnose across two
/// platforms. The nil sentinel accepts the first frame whatever its value and
/// enforces strict increase from then on.
struct MediaSequenceGate: Equatable {
    private(set) var lastAccepted: UInt64?

    init() {}

    /// Returns false if the frame must be rejected (and the session closed).
    mutating func admit(_ sequence: UInt64) -> Bool {
        if let last = lastAccepted, sequence <= last { return false }
        lastAccepted = sequence
        return true
    }
}

File: src/windows-native/LanMessenger/Core/Networking/Media/MediaReassembler.cs
namespace LanMessenger.Core.Networking.Media;

// The demux: per-channel fragment buffers, so an interleaved control or input
// frame arriving between two video segments never disturbs the partially
// assembled video frame. That per-channel keying is the reader-side half of the
// writer's interleaving design — without it, interleaving would corrupt video.
// Mirror of the macOS MediaReassembler.swift.

/// <summary>A fully reassembled inbound frame.</summary>
public readonly record struct MediaInboundFrame(
    MediaChannel Channel,
    MediaFlags Flags,
    /// Sequence of the FINAL fragment.
    ulong Sequence,
    /// Capture timestamp from the FIRST fragment.
    ulong CaptureUs,
    byte[] Payload,
    int FragmentCount);

public enum MediaReassemblyKind { Complete, Partial, Fault }

public readonly record struct MediaReassemblyOutcome(
    MediaReassemblyKind Kind,
    MediaInboundFrame Frame = default,
    MediaFaultKind Fault = default,
    string FaultMessage = "")
{
    public static MediaReassemblyOutcome Complete(MediaInboundFrame f) =>
        new(MediaReassemblyKind.Complete, f);
    public static MediaReassemblyOutcome Partial => new(MediaReassemblyKind.Partial);
    public static MediaReassemblyOutcome FaultWith(MediaFaultKind kind, string message) =>
        new(MediaReassemblyKind.Fault, default, kind, message);
}

public sealed class MediaReassembler
{
    private sealed class Pending
    {
        public readonly List<byte> Payload = [];
        public ulong CaptureUs;
        public MediaFlags Flags;
        public int FragmentCount;
    }

    private readonly Dictionary<byte, Pending> _pending = [];
    private readonly int _maxAssembledBytes;
    private readonly int _maxFragments;

    /// <summary>Two caps, not one, and they close different holes.</summary>
    /// <remarks>
    /// The 4 MiB wire cap bounds a single *frame*; it says nothing about the
    /// reassembled total, so the byte cap is needed to bound a long fragment
    /// chain. But a byte cap alone still lets a stream of zero-length fragments
    /// grow a per-channel buffer forever while every individual frame passes the
    /// wire check — a real allocation vector — so the fragment count is capped
    /// independently.
    /// </remarks>
    public MediaReassembler(
        int maxAssembledBytes = MediaFrameCodec.MaxFrameLength, int maxFragments = 512)
    {
        _maxAssembledBytes = maxAssembledBytes;
        _maxFragments = maxFragments;
    }

    public MediaReassemblyOutcome Accept(MediaFrameHeader header, byte[] plaintext)
    {
        if (header.Channel is not { } channel) return MediaReassemblyOutcome.Partial;
        byte key = header.RawChannel;

        // Not fragmented: a whole frame in one go.
        if (!header.IsFragmented)
        {
            if (_pending.ContainsKey(key))
            {
                // The writer never does this, so it is corruption or an attack.
                _pending.Remove(key);
                return MediaReassemblyOutcome.FaultWith(MediaFaultKind.WholeFrameMidReassembly,
                    $"unfragmented frame on channel {channel} with a partial frame pending");
            }
            return MediaReassemblyOutcome.Complete(new MediaInboundFrame(
                channel, header.Flags, header.Sequence, header.CaptureUs, plaintext, 1));
        }

        if (!_pending.TryGetValue(key, out var entry))
        {
            entry = new Pending { CaptureUs = header.CaptureUs, Flags = header.Flags };
            _pending[key] = entry;
        }

        if (entry.FragmentCount + 1 > _maxFragments)
        {
            _pending.Remove(key);
            return MediaReassemblyOutcome.FaultWith(MediaFaultKind.FragmentCountExceeded,
                $"reassembly on channel {channel} exceeded {entry.FragmentCount + 1} fragments");
        }
        if (entry.Payload.Count + plaintext.Length > _maxAssembledBytes)
        {
            _pending.Remove(key);
            return MediaReassemblyOutcome.FaultWith(MediaFaultKind.ReassemblyOverflow,
                $"reassembly on channel {channel} exceeded {entry.Payload.Count + plaintext.Length} bytes");
        }

        entry.Payload.AddRange(plaintext);
        entry.FragmentCount++;

        if (!header.IsFinalFragment) return MediaReassemblyOutcome.Partial;

        _pending.Remove(key);
        return MediaReassemblyOutcome.Complete(new MediaInboundFrame(
            channel,
            entry.Flags,          // first fragment's flags, incl. keyframe
            header.Sequence,      // final fragment's sequence
            entry.CaptureUs,      // first fragment's timestamp
            [.. entry.Payload],
            entry.FragmentCount));
    }

    /// <summary>Bytes currently held for a channel. Tests and the stats summary only.</summary>
    public int PendingBytes(MediaChannel channel) =>
        _pending.TryGetValue((byte)channel, out var e) ? e.Payload.Count : 0;

    public void Reset() => _pending.Clear();
}

/// <summary>Strictly-increasing sequence enforcement.</summary>
/// <remarks>
/// LastAccepted starts null rather than 0, and that choice is load-bearing. The
/// protocol says a reconnect restarts sequencing from zero against a key that has
/// never been used, so the first frame of a session legitimately carries sequence
/// 0. Initialising to 0 and requiring "> last" would reject it, and the session
/// would die at hello with a sequence violation that looks exactly like a crypto
/// fault — a miserable thing to diagnose across two platforms. The null sentinel
/// accepts the first frame whatever its value and enforces strict increase from
/// then on.
/// </remarks>
public struct MediaSequenceGate
{
    public ulong? LastAccepted { get; private set; }

    /// <summary>Returns false if the frame must be rejected (and the session closed).</summary>
    public bool Admit(ulong sequence)
    {
        if (LastAccepted is { } last && sequence <= last) return false;
        LastAccepted = sequence;
        return true;
    }
}
